cvmfs/resolv_conf_event_handler.cc

```diff
@@ -0,0 +1,85 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#include "resolv_conf_event_handler.h"
+
+#include <string>
+#include <vector>
+
+#include "backoff.h"
+#include "logging.h"
+#include "util/string.h"
+
+/*
+ * Returns any  nameserver IP addresses from file
+ *
+ * Reads a file, line by line. Returns the IP addresses corresponding
+ * to nameservers (i.e. from lines "nameserver <IP_ADDRESS>")
+ */
+bool GetDnsAddresses(const std::string& resolv_file,
+                     std::vector<std::string>* ipv4_addresses,
+                     std::vector<std::string>* ipv6_addresses) {
+  bool done = false;
+  BackoffThrottle throttle(100, 1000, 5000);
+  while (!done) {
+    FILE* f = std::fopen(resolv_file.c_str(), "r");
+    if (!f) {
+      LogCvmfs(kLogCvmfs, kLogDebug,
+               "ResolvConfEventHandler - Could not open /etc/resolv.conf");
+      throttle.Throttle();
+      continue;
+    }
+    std::string line;
+    while (GetLineFile(f, &line)) {
+      std::vector<std::string> key_val = SplitString(line, ' ');
+      if (key_val.size() == 2 && key_val[0] == "nameserver") {
+        // We have found a nameserver line
+        if (SplitString(key_val[1], '.').size() == 4) {
+          if (ipv4_addresses) {
+            // This looks like an IPv4 address
+            ipv4_addresses->push_back(key_val[1]);
+          }
+        } else if (SplitString(key_val[1], ':').size() == 6) {
+          if (ipv6_addresses) {
+            // This looks like an IPv6 address
+            ipv6_addresses->push_back(key_val[1]);
+          }
+        }
+      }
+    }
+    std::fclose(f);
+    done = true;
+  }
+  return true;
+}
+
+ResolvConfEventHandler::ResolvConfEventHandler(
+  download::DownloadManager* download_manager,
+  download::DownloadManager* external_download_manager)
+    : download_manager_(download_manager)
+    , external_download_manager_(external_download_manager) {}
+
+ResolvConfEventHandler::~ResolvConfEventHandler() {}
+
+bool ResolvConfEventHandler::Handle(const std::string& file_path,
+                                    file_watcher::Event /*event*/,
+                                    bool* clear_handler) {
+  std::vector<std::string> addresses;
+  GetDnsAddresses(file_path, &addresses, NULL);
```

Comment:
We might need a bit more logic here.  Users can set whether they want to prefer IPv4 or IPv6 through `CVMFS_IPFAMILY_PREFER`.  We should ask the download manager and, depending on the result, if there are IPv4 and IPv6 nameservers select accordingly.